Add 3Box profile integration to the Snapshot app

ChaituVR opened the pull request:
Changes proposed in this pull request:
- Add 3Box profile integration to the Snapshot app

### 1. Add profile in the top nav
![image](https://user-images.githubusercontent.com/15967809/100859917-dc056c80-34b5-11eb-98d4-fc16894d979f.png)

### 2. Add profile and button in account modal
If the User already have an account in 3Box, providing a link to edit
![image](https://user-images.githubusercontent.com/15967809/100859996-f2132d00-34b5-11eb-8b52-639f834e2f71.png)

if the User is not having an account in 3Box, providing a like to 3Box,
![image](https://user-images.githubusercontent.com/15967809/100860132-1cfd8100-34b6-11eb-818c-0514c1c49107.png)


### 3. Add names and avatars of the voters
URLs to test:
http://localhost:8080/#/balancer/proposal/QmQaxYxNp2Vq9PqW7Xc1Acnup7wC2Yrhg3UiYi9DYBFrPZ
More Votes: http://localhost:8080/#/yearn/proposal/QmUAUtyK7FiEAKF3FCeniWsvSiVxT2J6DESVcBKZYnPrCi

![image](https://user-images.githubusercontent.com/15967809/100860713-de1bfb00-34b6-11eb-86a2-857b0e4fb825.png)

### 4. Add name and avatar of the proposal author
Inside Proposal: (http://localhost:8080/#/balancer/proposal/QmQaxYxNp2Vq9PqW7Xc1Acnup7wC2Yrhg3UiYi9DYBFrPZ)
![image](https://user-images.githubusercontent.com/15967809/100860208-38688c00-34b6-11eb-9948-301b2af83d18.png)

### 5. Add profile in user modal
![image](https://user-images.githubusercontent.com/15967809/100861008-423ebf00-34b7-11eb-8c10-e81b96623dba.png)

bonustrack commented on src/components/Modal/Account.vue:
```diff
@@ -48,12 +48,40 @@
+          class="mb-2 d-block"
+        >
+          <UiButton class="button-outline width-full">
+            Edit Profile on 3Box
```
This should use "Sentence case"
```suggestion
            Edit profile on 3Box
```

bonustrack commented on src/components/Modal/Account.vue:
```diff
@@ -48,12 +48,40 @@
+          class="mb-2 d-block"
+        >
+          <UiButton class="button-outline width-full">
+            Connect Profile on 3Box
```
This button should have this label
```suggestion
            Create profile on 3Box
```

bonustrack commented on package.json:
```diff
@@ -9,6 +9,7 @@
     "lint": "vue-cli-service lint"
   },
   "dependencies": {
+    "3box": "^1.22.2",
```
You should commit the changes in package-lock.json too

bonustrack commented on src/components/Modal/User.vue:
```diff
@@ -1,8 +1,9 @@
-      <h3 v-text="_shorten(address)" />
+      <Avatar :profile="profile" :address="address" size="64" class="mb-4" />
+      <h3 v-if="profile && profile.name" class="mt-3" v-text="profile.name" />
+      <h3 v-else v-text="_shorten(address)" />
```
Space is missing between the avatar and h3 when there is no profile
```suggestion
      <h3 v-else v-text="_shorten(address)" class="mt-3" />
```

bonustrack commented on src/components/Avatar.vue:
```diff
@@ -1,13 +1,33 @@
+    <img
+      v-if="profile && profile.image"
+      :src="
+        `https://ipfs.io/ipfs/${
```
This should use the mixin `_ipfsUrl`

bonustrack commented on src/store/modules/web3.ts:
```diff
@@ -100,9 +102,11 @@ const actions = {
       } catch (e) {
         console.error(e);
       }
+      const profile = await Box.getProfile(account);
```
This can be loaded in the same time than `lookupAddress` on line 101 using `Promise.all()`. That would make login resolve faster.

bonustrack reviewed (commented):
Thank you for your contribution @ChaituVR. I've added some suggestions.

ChaituVR commented:
Thanks for the review @bonustrack Added those suggested changes,

also added `VUE_APP_IPFS_NODE=ipfs.io` to .env to make it work, let me know if it is not required
and also let me know if any other changes required for this feature

bonustrack commented on .env:
```diff
@@ -1,2 +1,3 @@
 VUE_APP_HUB_URL=https://testnet.snapshot.page
 VUE_APP_IPFS_GATEWAY=gateway.ipfs.io
+VUE_APP_IPFS_NODE=ipfs.io
```
You don't need this, there is already `VUE_APP_IPFS_GATEWAY` that does the same, i just noticed that it's not used in the mixin, could you change the config var there?

bonustrack commented on src/store/modules/app.ts:
```diff
@@ -154,12 +155,29 @@ const actions = {
         ipfsGet(gateway, id),
         client.request(`${space.key}/proposal/${id}`)
       ]);
+      const profile = await Box.getProfile(proposal.address)
```
It will be faster to resolve if you load the author profile within the Box.profileGraphQL request, we would have a single request instead of 2.

bonustrack reviewed (commented):
The size of the app bundle has been increased by x6 with the adding of 3box have a look here:
Without 3Box: https://snapshot.page/report.html
With 3Box: https://bafybeidq45eittd3vfbbydxd3fcwfwe47xvx3p5xbe7gimyztve44totx4.on.fleek.co/report.html
Is there a way to import only what we need from the dependency? I believe what we need is just to query the 3Box API endpoint, we don't need all the advanced functions.

ChaituVR commented:
Awesome suggestions :) Thanks for those.. fixed them and now the bundle size is 
![image](https://user-images.githubusercontent.com/15967809/101018981-47bd0780-3592-11eb-8606-9345609edc23.png)
also remove that getprofile call everywhere it is now using 3box's graphql only

bonustrack reviewed (commented):
Ok last changes, 
- Can you create a file helpers/3box.ts with a method `getProfiles` and use this method on app.ts and web3.ts, that will reduce duplicate code and make the function more accessible, you don't need to have 2 sub request on for voterProfiles and authorProfile, you could just request profiles of an array of addresses that include both author and voters addresses. 
- Can you resolve conflicts and and run the "lint" script?

ChaituVR commented:
@bonustrack already tried using `getProfiles` in the start itself but there is a cors issue with that function 

there is an open issue here, https://github.com/3box/3box-js/issues/649

so went with graphql method so we can get everything in one api call..

bonustrack commented:
Sorry i found 2 more issues:

1: When i connect my wallet here https://bafybeidqtbuxmtfemil2qzio3wf5zlonbcerbv2f2t73s3646yuhue7kim.on.fleek.co then refresh the page, the wallet should be reconnected again and there should be a loading indicator in top nav like this: 
![image](https://user-images.githubusercontent.com/16245250/101273403-ac14dc80-37c7-11eb-857b-b4c46f3bf3b3.png)
But currently there is no loading indicator.

2: The size of the dependency 3box is too heavy even using only `3box/lib/api`. It would be better if we just send a request to 3box API directly using the method `subgraphRequest`, this method is used in others place in Snapshot.js it's a standard way to query GraphQL APIs, and it's super light.

ChaituVR commented:
@bonustrack

Fixed that loading issue, that menu button is missing `app.authLoading`
```
<UiButton
      v-if="!$auth.isAuthenticated"
      @click="modalOpen = true"
      :loading="loading || app.authLoading"
>
```

and removed 3box's library didn't know graphql method exists for this :)
![image](https://user-images.githubusercontent.com/15967809/101279539-ecd11d80-37e8-11eb-8e67-ae2cada5b3d0.png)

bonustrack commented:
Thanks! That's much better now, i've just noticed there is an error when i load the page with a call to `profile.name` failing:
![image](https://user-images.githubusercontent.com/16245250/101280777-915d5a80-37fd-11eb-9121-60291961acaa.png)

bonustrack commented on src/components/Topnav.vue:
```diff
@@ -41,18 +41,28 @@
                 />
-                <span v-if="web3.name" v-text="web3.name" class="hide-sm" />
+                <span
+                  v-if="web3.profile.name"
```
```suggestion
                  v-if="web3.profile && web3.profile.name"
```

bonustrack reviewed (approved):
Great work @ChaituVR! The PR is good to be merged
